src/NuGet.Core/NuGet.Packaging/Signing/TrustStore/X509TrustStore.cs

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using NuGet.Common;
+
+namespace NuGet.Packaging.Signing
+{
+    /// <summary>
+    /// Represents the X.509 trust store that package signing and signed package verification will use.
+    /// </summary>
+    public static class X509TrustStore
+    {
+        private static IX509ChainFactory Instance;
+        private static readonly object LockObject = new();
+
+        internal static bool IsEnabled { get; }
+
+        static X509TrustStore()
+        {
+            Version DotNet7 = new(major: 7, minor: 0);
+
+            IsEnabled = Environment.Version >= DotNet7;
+        }
+
+        /// <summary>
+        /// Attempts to discover and load an X.509 trust store and log details about the attempt.
+        /// If initialization has already happened, a call to this method will have no effect.
+        /// </summary>
+        /// <remarks>Explicit initialization is unnecessary for X.509 certificate chain building;
+        /// initialization will happen automatically on demand.  This method provides an opportunity
+        /// to log diagnostic information about trust store discovery.</remarks>
+        /// <param name="logger">A logger.</param>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="logger" /> is <c>null</c>.</exception>
+        public static void Initialize(ILogger logger)
+        {
+            if (logger is null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            _ = GetX509ChainFactory(logger);
+        }
+
+        internal static IX509ChainFactory GetX509ChainFactory(ILogger logger)
+        {
+            if (logger is null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            if (Instance is not null)
+            {
+                return Instance;
+            }
+
+            lock (LockObject)
+            {
+                if (Instance is not null)
+                {
+                    return Instance;
+                }
+
+                Instance = CreateX509ChainFactory(logger);
+            }
+
+            return Instance;
+        }
+
+        // Non-private for testing purposes only
+        internal static IX509ChainFactory CreateX509ChainFactory(ILogger logger)
+        {
+            if (IsEnabled)
+            {
+#if NET5_0_OR_GREATER
+                if (RuntimeEnvironmentHelper.IsLinux)
+                {
+                    if (SystemCertificateBundleX509ChainFactory.TryCreate(
```

Comment:
How long do these initializations take?

Reply:
I measured [`X509TrustStore.Initialize(...)`](https://github.com/NuGet/NuGet.Client/pull/4585/files#diff-803284c550d9fa1b2d6f81d0fa10742c06e2dac0db399f9cb2c8178c69ebd09eR93).

On an Intel Core i7-7500U CPU @ 2.70GHz running Ubuntu 18.04.6 x64 with .NET SDK 7.0.100-preview.4.22252.9, I measured 10 iterations:

| Elapsed (ms) |
| -- |
| 29.6712 |
| 30.1272 |
| 29.432 |
| 29.9885 |
| 29.2739 |
| 31.6841 |
| 29.3276 |
| 37.9535 |
| 29.486 |
| 29.562 |

Average:  30.7 ms
95th percentile:  35.1 ms

The same test on an Intel E5-2673 v3 CPU @ 2.40GHz running Windows 10.0.19044 x64:

| Elapsed (ms) |
| -- |
| 1.8628 |
| 2.0156 |
| 2.8631 |
| 1.8689 |
| 2.9112 |
| 1.8395 |
| 3.0992 |
| 1.8976 |
| 2.1964 |
| 5.2473 |

Average:  2.6 ms
95th percentile:  4.3 ms

Keep in mind that these are one-time costs.  The scenario is slower on Linux because (of course) we have to load a root certificate bundle from disk.

Note:  I realized later that I had used a Debug configuration build of NuGet assemblies.

Reply:
Thanks!

Surprised it even registers on windows. Should we have a env check earlier? 
I'm really not suggesting a change or anything like that.

> Note: I realized later that I had used a Debug configuration build of NuGet assemblies.

Even better then :D Release is faster